src/goto-symex/goto_symex_state.cpp

```diff
@@ -591,68 +598,144 @@ void goto_symex_statet::rename_address(exprt &expr, const namespacet &ns)
+
+      if(renamed_array.has_value())
+      {
+        result.array() = std::move(*renamed_array);
+        result.type() = to_array_type(result.array().type()).subtype();
+      }
 
-      if_expr.type()=if_expr.true_case().type();
+      if(renamed_index != index_expr.index())
+        result.index() = std::move(renamed_index);
+
+      return std::move(result);
     }
-    else if(expr.id()==ID_member)
+    else
+      return {};
+  }
+  else if(expr.id() == ID_if)
+  {
+    // the condition is not an address
+    const if_exprt &if_expr = to_if_expr(expr);
+    auto renamed_cond = rename<level>(if_expr.cond(), ns);
+    auto renamed_true = rename_address<level>(if_expr.true_case(), ns);
+    auto renamed_false = rename_address<level>(if_expr.false_case(), ns);
+
+    if(
+      renamed_cond != if_expr.cond() || renamed_true.has_value() ||
+      renamed_false.has_value())
     {
-      member_exprt &member_expr=to_member_expr(expr);
+      if_exprt result = if_expr;
 
-      rename_address<level>(member_expr.struct_op(), ns);
+      if(renamed_cond != if_expr.cond())
+        result.cond() = std::move(renamed_cond);
 
-      // type might not have been renamed in case of nesting of
-      // structs and pointers/arrays
-      if(
-        member_expr.struct_op().type().id() != ID_symbol_type &&
-        member_expr.struct_op().type().id() != ID_struct_tag &&
-        member_expr.struct_op().type().id() != ID_union_tag)
+      if(renamed_true.has_value())
       {
-        const struct_union_typet &su_type=
-          to_struct_union_type(member_expr.struct_op().type());
-        const struct_union_typet::componentt &comp=
-          su_type.get_component(member_expr.get_component_name());
-        PRECONDITION(comp.is_not_nil());
-        expr.type()=comp.type();
+        result.true_case() = std::move(*renamed_true);
+        result.type() = result.true_case().type();
       }
-      else
-        rename<level>(expr.type(), irep_idt(), ns);
+
+      if(renamed_false.has_value())
+        result.false_case() = std::move(*renamed_false);
+
+      return std::move(result);
+    }
+    else
+      return {};
+  }
+  else if(expr.id() == ID_member)
+  {
+    const member_exprt &member_expr = to_member_expr(expr);
+
+    auto renamed_struct_op = rename_address<level>(member_expr.struct_op(), ns);
+
+    // type might not have been renamed in case of nesting of
+    // structs and pointers/arrays
+    optionalt<typet> renamed_type;
+    if(
+      renamed_struct_op.has_value() &&
+      renamed_struct_op->type().id() != ID_symbol_type &&
+      renamed_struct_op->type().id() != ID_struct_tag &&
+      renamed_struct_op->type().id() != ID_union_tag)
+    {
+      const struct_union_typet &su_type =
+        to_struct_union_type(renamed_struct_op->type());
+      const struct_union_typet::componentt &comp =
+        su_type.get_component(member_expr.get_component_name());
+      PRECONDITION(comp.is_not_nil());
```

Comment:
I'm not sure this really qualifies as a precondition

Reply:
True, but also this was just what it was before (this code is now just indented one level less). Nevertheless I've changed into into a `DATA_INVARIANT`.